### PR TITLE
refactor!: implement AxelarExecutable and AxelarExecutableWithToken in ITS

### PR DESCRIPTION
**Pull request by ahramy**
[AXE-4660](https://axelarnetwork.atlassian.net/browse/AXE-4660) : Inherit the simpler [AxelarGMPExecutableWithToken](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/main/contracts/executable/AxelarGMPExecutableWithToken.sol) abstract contract in ITS contract instead of [re-implementing](https://github.com/axelarnetwork/interchain-token-service/blob/main/contracts/InterchainTokenService.sol#L689) it. The original AxelarExecutable wasn’t including the commandId and it was including the full gateway interface that’s susceptible to changes.

[AXE-5795](https://axelarnetwork.atlassian.net/browse/AXE-5795)

[AXE-4660]: https://axelarnetwork.atlassian.net/browse/AXE-4660?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[AXE-5795]: https://axelarnetwork.atlassian.net/browse/AXE-5795?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/InterchainTokenService.sol [hunk 1]
```diff
@@ -871,6 +845,16 @@ contract InterchainTokenService is
...
     }
 
+    function callContractWithToken(
+        string calldata destinationChain,
+        string calldata contractAddress,
+        bytes calldata payload,
+        string calldata symbol,
+        uint256 amount
+    ) external override {
+        // do nothing ?
```
Comment (ahramy) [inline contracts/InterchainTokenService.sol:855]: I'm not sure what to do with the `callContractWithToken` function here because it's already implemented as `_callContractWithToken` and is called within `_transmitInterchainTransfer` and `interchainTransfer` (tightly coupled with the `metadata`). Could you give me some guidance?

File: contracts/InterchainTokenFactory.sol [hunk 2]
```diff
@@ -51,6 +51,9 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
         return CONTRACT_ID;
     }
 
+    function _setup(bytes calldata params) internal override {
```
[the remark this answers is not in the document]
Comment (milapsheth) [reply contracts/InterchainTokenFactory.sol:54]: we probably marked this as abstract in a recent release making this required, so yeah. you can see the history for it

File: contracts/InterchainTokenService.sol (hunk 1 above)
Comment (ahramy) [reply contracts/InterchainTokenService.sol]: had a call and discussed.

**Comment by ahramy** [comment]
One Unit test is failing, which is related to the upgrade flow + governance logic 

```
  238 passing (5s)
  1 failing

  1) Interchain Token Service Upgrade Flow
       "before all" hook for "should upgrade Interchain Token Service through AxelarServiceGovernance timeLock proposal":
     Error: invalid address or ENS name (argument="name", value=["0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266","0x70997970C51812dc3A010C7d01b50e0d17dc79C8","0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"], code=INVALID_ARGUMENT, version=contracts/5.7.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at Logger.throwArgumentError (node_modules/@ethersproject/logger/src.ts/index.ts:285:21)
      at /Users/zenik/research/interchain-token-service/node_modules/@ethersproject/contracts/src.ts/index.ts:123:16
      at step (node_modules/@ethersproject/contracts/lib/index.js:48:23)
      at Object.next (node_modules/@ethersproject/contracts/lib/index.js:29:53)
      at fulfilled (node_modules/@ethersproject/contracts/lib/index.js:20:58)
```

and I found that the test is calling [AxelarServiceGovernance](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/6a86f3bc8002ed6c0c7cd952e93d80ea4018c1cc/contracts/governance/AxelarServiceGovernance.sol#L13
) and the [IInterchainGovernance](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/6a86f3bc8002ed6c0c7cd952e93d80ea4018c1cc/contracts/interfaces/IInterchainGovernance.sol#L13) is using `IAxelarExecutable`. I  changed it to `IAxelarGMPExecutableWithToken`, it's a huge change.. seems like require to change a lot of codes.

Could you review what I’ve done? I think we need to discuss further how we handle the Governance upgrade.

File: contracts/InterchainTokenFactory.sol [hunk 3]
```diff
@@ -40,7 +39,7 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
         interchainTokenService = IInterchainTokenService(interchainTokenService_);
 
         chainNameHash = interchainTokenService.chainNameHash();
-        gateway = interchainTokenService.gateway();
+        gateway = interchainTokenService.getGatewayWithToken();
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:42]:
```suggestion
        gateway = IAxelarGMPGatewayWithToken(interchainTokenService.gateway());
```

File: contracts/InterchainTokenService.sol [hunk 4]
```diff
@@ -377,13 +376,13 @@ contract InterchainTokenService is
...
         bytes calldata payload
-    ) public payable whenNotPaused {
+    ) public payable override whenNotPaused {
         uint256 messageType = abi.decode(payload, (uint256));
         if (messageType != MESSAGE_TYPE_INTERCHAIN_TRANSFER) {
             revert InvalidExpressMessageType(messageType);
         }
 
-        if (gateway.isCommandExecuted(commandId)) revert AlreadyExecuted();
+        if (gatewayWithToken().isCommandExecuted(commandId)) revert AlreadyExecuted();
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:385]:
```suggestion
        if (gateway().isCommandExecuted(commandId)) revert AlreadyExecuted();
```
this should still work. only use `gatewayWithToken` when that functionality is needed

File: contracts/InterchainTokenService.sol [hunk 5]
```diff
@@ -901,20 +869,21 @@ contract InterchainTokenService is
         bytes32 commandId,
         string calldata sourceChain,
         string calldata sourceAddress,
-        bytes memory payload,
-        bytes32 payloadHash
-    ) internal {
+        bytes calldata payload
+    ) internal override whenNotPaused {
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:873]:
```suggestion
    ) internal override onlyRemoteService(sourceChain, sourceAddress) whenNotPaused {
```
missing modifier from `execute`. Same for `_executeWithToken`

File: contracts/interfaces/IInterchainTokenService.sol [hunk 6]
```diff
@@ -94,6 +95,12 @@ interface IInterchainTokenService is
     );
     event InterchainTokenIdClaimed(bytes32 indexed tokenId, address indexed deployer, bytes32 indexed salt);
 
+    /**
+     * @notice Returns the address of the IAxelarGMPGatewayWithToken contract.
+     * @return The Axelar GMP Gateway with Token instance.
+     */
+    function getGatewayWithToken() external view returns (IAxelarGMPGatewayWithToken);
```
Comment (milapsheth) [inline contracts/interfaces/IInterchainTokenService.sol:102]: not needed

File: contracts/InterchainTokenService.sol [hunk 7]
```diff
@@ -622,26 +621,6 @@ contract InterchainTokenService is
...
-        bytes32 commandId,
-        string calldata sourceChain,
-        string calldata sourceAddress,
-        bytes calldata payload
-    ) external onlyRemoteService(sourceChain, sourceAddress) whenNotPaused {
-        bytes32 payloadHash = keccak256(payload);
-
-        if (!gateway.validateContractCall(commandId, sourceChain, sourceAddress, payloadHash)) revert NotApprovedByGateway();
-
-        _execute(commandId, sourceChain, sourceAddress, payload, payloadHash);
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:642]: To avoid paying the gas cost for computing `payloadHash` again, instead of inheriting the executable, we'd need to keep this implementation

File: contracts/InterchainTokenService.sol [hunk 8]
```diff
@@ -901,20 +869,21 @@ contract InterchainTokenService is
         bytes32 commandId,
         string calldata sourceChain,
         string calldata sourceAddress,
-        bytes memory payload,
-        bytes32 payloadHash
-    ) internal {
+        bytes calldata payload
+    ) internal override whenNotPaused {
+        bytes32 payloadHash = keccak256(payload);
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:874]: the correct fix for this needs to be to inherit from a modified version of ValuedExpressExecutable which can handle the express logic. That's more work for this PR. So, let's keep the current `execute`/`executeWithToken` implementation we're using for now

File: package.json [hunk 9]
```diff
@@ -29,8 +29,11 @@
     "node": ">=18"
   },
   "dependencies": {
-    "@axelar-network/axelar-cgp-solidity": "6.2.1",
-    "@axelar-network/axelar-gmp-sdk-solidity": "5.6.4"
+    "@axelar-network/axelar-cgp-solidity": "file:../axelar-cgp-solidity",
```
Comment (milapsheth) [inline package.json:32]: this hasn't changed, should point to the latest release we have

File: package.json [hunk 10]
```diff
@@ -29,8 +29,11 @@
...
+    "@axelar-network/axelar-gmp-sdk-solidity": "file:../axelar-gmp-sdk-solidity",
+    "build": "^0.1.4",
+    "npm": "^10.8.3",
+    "run": "^1.5.0"
```
Comment (milapsheth) [inline package.json:36]: what are these?

File: contracts/InterchainTokenService.sol [hunk 11]
```diff
@@ -4,14 +4,14 @@ pragma solidity ^0.8.0;
...
-import { IAxelarGateway } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IAxelarGateway.sol';
+import { IAxelarGMPGatewayWithToken } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IAxelarGMPGatewayWithToken.sol';
+import { AxelarGMPExecutableWithToken } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/executable/AxelarGMPExecutableWithToken.sol';
 import { ExpressExecutorTracker } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/express/ExpressExecutorTracker.sol';
 import { Upgradable } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/upgradable/Upgradable.sol';
 import { AddressBytes } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/libs/AddressBytes.sol';
 import { Multicall } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/utils/Multicall.sol';
 import { Pausable } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/utils/Pausable.sol';
 import { InterchainAddressTracker } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/utils/InterchainAddressTracker.sol';
-
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:14]: keep the newline to separate external/internal imports

File: contracts/InterchainTokenFactory.sol (hunk 3 above)
Comment (milapsheth) [reply contracts/InterchainTokenFactory.sol:42]: this is going to be removed anyways since gateway registration needs to be dropped

File: contracts/InterchainTokenService.sol [hunk 12]
```diff
@@ -901,20 +869,21 @@ contract InterchainTokenService is
...
-        bytes memory payload,
-        bytes32 payloadHash
-    ) internal {
+        bytes calldata payload
+    ) internal override whenNotPaused {
+        bytes32 payloadHash = keccak256(payload);
         uint256 messageType;
         string memory originalSourceChain;
-        (messageType, originalSourceChain, payload) = _getExecuteParams(sourceChain, payload);
+        bytes memory originalPayload;
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:877]: this is not storing the original payload. this is the transformed payload. although, reverting the executable inheritance should revert the change here anyways

Comment (Foivos) [reply contracts/InterchainTokenService.sol:877]: also, what is the point of introducing the extra local variable? The payload that is preserved does not seem to be used anywhere.

File: contracts/InterchainTokenService.sol (hunk 5 above)
Comment (ahramy) [reply contracts/InterchainTokenService.sol:873]: Nice catch, updated.

File: contracts/InterchainTokenService.sol [hunk 13]
```diff
@@ -44,7 +44,7 @@ contract InterchainTokenService is
     using AddressBytes for bytes;
     using AddressBytes for address;
 
-    IAxelarGateway public immutable gateway;
+    IAxelarGMPGatewayWithToken public immutable gatewayAddress;
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:47]:
```suggestion
    IAxelarGMPGateway public immutable gateway;
```
I'd rather keep this pure GMP, and explicitly cast to GMP with token when needed, since that's legacy functionality

Add comments around pure gmp amplifier vs gmp with token legacy gateway support for ITS (latter is expected to revert and only work on chains using Axelar consensus connection)

File: contracts/utils/GatewayCaller.sol [hunk 14]
```diff
@@ -3,15 +3,15 @@
...
+import { IAxelarGMPGatewayWithToken } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IAxelarGMPGatewayWithToken.sol';
 import { IGatewayCaller } from '../interfaces/IGatewayCaller.sol';
 
 /**
  * @title GatewayCaller contract
  * @dev This contract is used to handle cross-chain ITS calls via the Axelar gateway.
  */
 contract GatewayCaller is IGatewayCaller {
-    IAxelarGateway public immutable gateway;
+    IAxelarGMPGatewayWithToken public immutable gateway;
```
Comment (milapsheth) [inline contracts/utils/GatewayCaller.sol:14]: Same here, cast to legacy functionality when needed

File: test/InterchainTokenServiceUpgradeFlow.js [hunk 15]
```diff
@@ -19,7 +19,7 @@ const Create3Deployer = getContractJSON('Create3Deployer');
 const MINT_BURN = 4;
 
 describe('Interchain Token Service Upgrade Flow', () => {
-    let wallet, otherWallet, signer;
+    let wallet, otherWallet, multisig;
```
Comment (milapsheth) [inline test/InterchainTokenServiceUpgradeFlow.js:22]:
```suggestion
    let wallet, otherWallet, operator;
```
for clarity of the role. it's not a multisig, and the contract doesn't care about it

File: contracts/InterchainTokenService.sol (hunk 13 above)
Comment (ahramy) [reply contracts/InterchainTokenService.sol:47]: I see. Updated.

File: contracts/utils/GatewayCaller.sol (hunk 14 above)
Comment (ahramy) [reply contracts/utils/GatewayCaller.sol:14]: done.

File: test/InterchainTokenServiceUpgradeFlow.js (hunk 15 above)
Comment (ahramy) [reply test/InterchainTokenServiceUpgradeFlow.js:22]: I see. updated.

File: contracts/InterchainTokenService.sol [hunk 16]
```diff
@@ -707,8 +714,10 @@ contract InterchainTokenService is
     function _checkPayloadAgainstGatewayData(bytes memory payload, string calldata tokenSymbol, uint256 amount) internal view {
         (, bytes32 tokenId, , , uint256 amountInPayload) = abi.decode(payload, (uint256, bytes32, uint256, uint256, uint256));
 
-        if (validTokenAddress(tokenId) != gatewayAddress.tokenAddresses(tokenSymbol) || amount != amountInPayload)
-            revert InvalidGatewayTokenTransfer(tokenId, payload, tokenSymbol, amount);
+        if (
+            validTokenAddress(tokenId) != IAxelarGMPGatewayWithToken(address(gateway)).tokenAddresses(tokenSymbol) ||
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:718]:
```suggestion
        if (
            validTokenAddress(tokenId) != gatewayWithToken().tokenAddresses(tokenSymbol) ||
```
create an internal gatewayWithToken method since this needs to be done in more than one place (if it was a single one, then i would have recommended to create a variable for it)

File: contracts/InterchainTokenService.sol [hunk 17]
```diff
@@ -44,7 +44,14 @@ contract InterchainTokenService is
     using AddressBytes for bytes;
     using AddressBytes for address;
 
-    IAxelarGMPGatewayWithToken public immutable gatewayAddress;
+    /**
+     * @dev Axelar GMP Gateway for cross-chain messaging.
+     * Two available approaches:
+     * 1. Pure GMP: Uses IAxelarGMPGateway to handle messaging without token transfers, compatible across Amplifier chains.
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:50]:
```suggestion
     * 1. Cross-chain messaging (GMP): The Axelar Gateway allows sending cross-chain messages. This is compatible across both Amplifier and consensus chains. IAxelarGMPGateway interface exposes this functionality.
```

File: contracts/InterchainTokenService.sol [hunk 18]
```diff
@@ -44,7 +44,14 @@ contract InterchainTokenService is
     using AddressBytes for bytes;
     using AddressBytes for address;
 
-    IAxelarGMPGatewayWithToken public immutable gatewayAddress;
+    /**
+     * @dev Axelar GMP Gateway for cross-chain messaging.
+     * Two available approaches:
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:49]:
```suggestion
     * @dev There are two types of Axelar Gateways for cross-chain messaging:
```
ok to be a bit more wordy for docs

File: contracts/InterchainTokenService.sol [hunk 19]
```diff
@@ -44,7 +44,14 @@ contract InterchainTokenService is
...
+     * 1. Pure GMP: Uses IAxelarGMPGateway to handle messaging without token transfers, compatible across Amplifier chains.
+     * 2. GMP with Token: Legacy functionality. Explicitly cast to IAxelarGMPGatewayWithToken when needed.
+     *    This approach only functions on chains that rely on Axelar consensus Connection.
```
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:52]:
```suggestion
     * 2. Cross-chain messaging with Gateway Token: The AxelarGateway on legacy consensus EVM connections supports this (via callContractWithToken) but not Amplifier chains. `gateway` is cast to IAxelarGMPGatewayWithToken when gateway tokens need to be handled. ITS deployments on Amplifier chains will revert when this functionality is used.
```

File: contracts/InterchainTokenService.sol (hunk 16 above)
Comment (ahramy) [reply contracts/InterchainTokenService.sol:718]: Yeah, I agree with your approach. Updated.

File: contracts/InterchainTokenService.sol (hunk 18 above)
Comment (ahramy) [reply contracts/InterchainTokenService.sol:49]: I see. updated.

File: contracts/InterchainTokenService.sol (hunk 17 above)
Comment (ahramy) [reply contracts/InterchainTokenService.sol:50]: updated.

File: contracts/InterchainTokenService.sol (hunk 19 above)
Comment (ahramy) [reply contracts/InterchainTokenService.sol:52]: updated.

**Review by milapsheth — APPROVED** [review]

File: contracts/InterchainTokenService.sol [hunk 20]
```diff
@@ -377,7 +385,7 @@ contract InterchainTokenService is
         string calldata sourceChain,
         string calldata sourceAddress,
         bytes calldata payload
-    ) public payable whenNotPaused {
+    ) public payable override whenNotPaused {
```
Comment (hydrobeam) [inline contracts/InterchainTokenService.sol:388]: Just wondering why this function was marked `override`

File: test/InterchainTokenServiceUpgradeFlow.js [hunk 21]
```diff
@@ -86,14 +85,7 @@ describe('Interchain Token Service Upgrade Flow', () => {
...
         axelarServiceGovernance = await axelarServiceGovernanceFactory
-            .deploy(
-                gateway.address,
-                governanceChain,
-                governanceAddress,
-                buffer,
-                signers.map((signer) => signer.address),
-                threshold,
-            )
+            .deploy(gateway.address, governanceChain, governanceAddress, minimumTimeDelay, operator.address)
```
Comment (hydrobeam) [inline test/InterchainTokenServiceUpgradeFlow.js:88]: Why did the number of parameters being passed to `deploy` change - was `thresehold` not needed anymore?

**Review by milapsheth — APPROVED** [review]

File: test/InterchainToken.js [hunk 22]
```diff
@@ -147,7 +147,7 @@ describe('InterchainToken', () => {
             const contractBytecodeHash = keccak256(contractBytecode);
 
             const expected = {
-                london: '0xa01cf28b0b6ce6dc3b466e995585d69486400d671fce0ea8d06beba583e6f3bb',
+                london: '0x75a10f371014f97a894e69395422a36cc9e9961e32371c0213ed514702966b24',
```
Comment (milapsheth) [inline test/InterchainToken.js:150]: Why is this changing? These bytecode checks exist to catch unexpected modifications

**Review by milapsheth — CHANGES_REQUESTED** [review]

File: test/InterchainToken.js (hunk 22 above)
Comment (ahramy) [reply test/InterchainToken.js:150]: I flattened the `InterchainToken.sol` file and looked into the differences. I found that some functions were updated by this PR: https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/pull/189.

i.e.: _toAccountRoles
    `function _toAccountRoles(uint8[] memory roles) internal pure returns (uint256)  `
vs
   ` function _toAccountRoles(uint8[] memory roles) internal pure returns (uint256 accountRoles)`

Let me know if we're okay with this change or not.

Comment (milapsheth) [reply test/InterchainToken.js:150]: i see, let's revert it, and release 6.0.1

Comment (ahramy) [reply test/InterchainToken.js:150]: Merged PR: https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/pull/194 and release snapshot, and now the bytecode doesn't needs to be updated.

**Review by milapsheth — APPROVED** [review]
